### PR TITLE
nixos/pulseaudio: disable flat-volumes by default

### LINKED ISSUE #5632: PulseAudio: Disable flat volume?
Flat volume is enabled by default. I think we should disable it because this option could damage speakers/ears. What do you think?

Here is the relevant reddit post: http://www.reddit.com/r/linux/comments/2rjiaa/horrible_decisions_flat_volumes_in_pulseaudio_a/

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
Some applications are unaware of this feature and can set their volume to 100% on startup harming people ears and possiblly blowing someone's audio setup.

I noticed this in #54594 and by extension [epiphany](https://bugzilla.gnome.org/show_bug.cgi?id=675217
).

Please also note that many other distros have this default for the reason outlined above.

Closes #5632 #54594

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Infinisil** [comment]
I think this warrants a release notes entry mentioning what this effectively changes to volumes.

If I read correctly, `flat-volumes = no` means that when an application sets its volume, this number gets first scaled by your application-specific volume first. So if you have Firefox on 30%, and the application requests 50%, this will translate to 15% in the end. This in a way sandboxes every application into its own volume range.

`flat-volumes = yes` however makes this number go unscaled through pulseaudio. So when you have Firefox on 30%, and the application requests 50%, this will actually change your setting from 30% to 50%. If you need to keep your volume low because of your amplifier, this of course could blast your ears when it suddenly jumps to 100%.

Somebody correct me if I'm wrong.

File: nixos/modules/config/pulseaudio.nix [hunk 1]
```diff
@@ -178,9 +178,9 @@ in {
 
         config = mkOption {
           type = types.attrsOf types.unspecified;
-          default = {};
+          default = { flat-volumes = "no"; };
```
Comment (Infinisil) [inline nixos/modules/config/pulseaudio.nix]: If we set the default here, if the user sets the config on their own to something, this value gets overridden, which is probably not what people would expect. I think it should be set like this in the config section of this module instead:

```nix
{
  config = {
    hardware.pulseaudio.config.flat-volumes = mkDefault "no";
  };
}
```

Then people can set other values without having `flat-volumes` turned on again.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/config/pulseaudio.nix (hunk 1 above)
Comment (worldofpeace) [reply nixos/modules/config/pulseaudio.nix]: Yes that is correct :+1:

**Comment by worldofpeace** [comment]
> If I read correctly, flat-volumes = no means that when an application sets its volume, this number gets first scaled by your application-specific volume first. So if you have Firefox on 30%, and the application requests 50%, this will translate to 15% in the end. This in a way sandboxes every application into its own volume range.

> flat-volumes = yes however makes this number go unscaled through pulseaudio. So when you have Firefox on 30%, and the application requests 50%, this will actually change your setting from 30% to 50%. If you need to keep your volume low because of your amplifier, this of course could blast your ears when it suddenly jumps to 100%.

Disabling flat volumes enables relative volumes, apt to being relative to an absolute master volume.

> I think this warrants a release notes entry mentioning what this effectively changes to volumes.

Its effect is that we don't torture people anymore :rofl:

###### See: https://youtu.be/Skl1Gf8RCCw?t=54

I'll write a pretty terse one.

**Comment by worldofpeace** [comment]
For anyone unclear on volume model sitiuation, [this comment](https://bugzilla.redhat.com/show_bug.cgi?id=1265267#c3) could be helpful to read.

**Review by infinisil — APPROVED** [review]

**Comment by worldofpeace** [comment]
I think I'm going to merge this right away because the conclusion to do this has been already reached several times in other places. So we really don't need comuntity feedback for a solution that's already substantiated.

**Pull request merged**